File: MaximumProductSubarray.py
"""
152. Maximum Product Subarray
Find the contiguous subarray within an array (containing at least one number) which has the largest product.

For example, given the array [2,3,-2,4],
the contiguous subarray [2,3] has the largest product = 6.
"""
# O(n)
# O(1)
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        local_min, local_max, global_max = 1, 1, float("-inf")
        for num in nums:
            local_min, local_max = min(local_min * num, local_max * num, num), max(local_min * num, local_max * num, num)
            global_max = max(global_max, local_max)
        return global_max

if __name__ == "__main__":
    print(Solution().maxProduct([2, 3, -2, 4]))

File: Java/MaximumProductSubarray.java

public class MaximumProductSubarray {
    public int maxProduct(int[] nums) {
        int globalMax = nums[0];
        int localMax = nums[0];
        int localMin = nums[0];
        int tem;
        for(int i = 1; i < nums.length; i++) {
            tem = localMax;
            localMax = Math.max(localMax * nums[i], Math.max(localMin * nums[i], nums[i]));
            localMin = Math.min(tem * nums[i], Math.min(localMin * nums[i], nums[i]));
            globalMax = Math.max(globalMax, localMax);
        }
        return globalMax;
    }

}
